miniapps/neml2/constitutive.cpp

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2010-2025, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "constitutive.hpp"
+#include "neml2/tensors/functions/mv.h"
+#include <torch/torch.h>
+
+namespace op {
+
+static neml2::Tensor mfem_to_neml2_tensor(const neml2::TensorOptions &options,
+                                          ParameterFunction &pf) {
+  // Pointer to host/device data (MFEM owns it)
+  real_t *strain_data = pf.ReadWrite();
+
+  // Use torch::from_blob to create a tensor wrapper around the data
+  // (without copying!)
+  auto &q_space = pf.GetParameterSpace();
+  auto neml2_tensor = neml2::Tensor(
+      torch::from_blob(strain_data,
+                       {(int64_t)(q_space.GetTrueVSize() / q_space.GetVDim()),
+                        (int64_t)q_space.GetVDim()},
+                       options),
+      /*batch_dim=*/1);
+  return neml2_tensor;
+}
+
+static void neml2_to_mfem_tensor(neml2::Tensor &neml2_tensor,
+                                 ParameterFunction &pf) {
+  // Copy the stress data back to MFEM
+  // (either a host-host or device-device copy)
+  auto &q_space = pf.GetParameterSpace();
+  auto nbyte = q_space.GetTrueVSize() * sizeof(real_t);
+  if (neml2_tensor.options().device().is_cpu())
+    std::memcpy(pf.HostWrite(), neml2_tensor.data_ptr<real_t>(), nbyte);
+  else if (neml2_tensor.options().device().is_cuda())
+    MFEM_GPU_CHECK(cudaMemcpy(pf.ReadWrite(), neml2_tensor.data_ptr<real_t>(),
+                              nbyte, cudaMemcpyDeviceToDevice));
```

Comment:
I'm pretty sure MFEM provides device write that is backend agnostic. Maybe just `pf.Write()`?